File: app/src/main/java/com/donald/musictheoryapp/Question/IntervalInputQuestion.java
package com.donald.musictheoryapp.Question;

import com.donald.musictheoryapp.Music.MusicXML.Note;
import com.donald.musictheoryapp.Music.MusicXML.Score;
import com.donald.musictheoryapp.Utils.JSONArrayUtil;

import org.json.JSONException;
import org.json.JSONObject;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;

public class IntervalInputQuestion extends Question
{
	public static class Answer implements Question.Answer
	{
		public Note userAnswer;
		public Note correctAnswer;

		public Answer(Note userAnswer, Note correctAnswer)
		{
			this.userAnswer = userAnswer;
			this.correctAnswer = correctAnswer;
		}

		@Override
		public boolean correct()
		{
			return userAnswer.equals(correctAnswer);
		}

		public static Answer fromJSON(JSONObject object) throws JSONException
		{
			return new Answer(
				object.isNull("user_answer") ? null : Note.fromJSON(object.getJSONObject("user_answer")),
				Note.fromJSON(object.getJSONObject("correct_answer"))
			);
		}
	}

	public Answer answer;
	public Score score;
	public Note correctNote;
	public String requiredInterval;

	public IntervalInputQuestion() {}

	@Override
	public int points()
	{
		Note[] notes = score.parts()[0].measures()[0].notes();
		if(notes[1].equals(correctNote))
		{
			return 1;
		}
		else return 0;
	}

	public Score score() { return score; }
	public String requiredInterval() { return requiredInterval; }

	@Override
	public void acceptVisitor(QuestionVisitor visitor) { visitor.visit(this); }

	public static IntervalInputQuestion fromJSON(JSONObject object, QuestionGroup group)
		throws JSONException, IOException, XmlPullParserException
	{
		IntervalInputQuestion question = new IntervalInputQuestion();
		question.number = object.getInt("number");
		question.group = group;
		question.descriptions = JSONArrayUtil.descriptions(object);
		question.score = Score.fromXML(object.getString("score"));
		if(question.score.parts().length != 1 ||
			question.score.parts()[0].measures().length != 1 ||
			question.score.parts()[0].measures()[0].notes().length != 3)
		{
			throw new IllegalStateException("Invalid score from question.");
		}
		question.answer = Answer.fromJSON(object.getJSONObject("answer"));
		question.requiredInterval = object.getString("required_interval");
		/*
		return new IntervalInputQuestion(
			object.getInt("number"),
			group,
			JSONArrayUtil.descriptions(object),
			//Note.Pitch.fromJSON(object.getJSONObject("correct_pitch")),
			//Note.Pitch.fromJSON(object.getJSONObject("given_pitch")),
			Score.fromXML(object.getString("score")),
			Note.fromJSON(object.getJSONObject("correct_note")),
			object.getString("required_interval")
			//Clef.fromJSON(object.getJSONObject("clef")),
			//Key.fromJSON(object.getJSONObject("key"))
		);

		 */
		return question;
	}
}

File: app/src/main/java/com/donald/musictheoryapp/question/IntervalInputQuestion.kt
package com.donald.musictheoryapp.question

import kotlin.Throws
import com.donald.musictheoryapp.music.musicxml.Score
import org.xmlpull.v1.XmlPullParserException
import com.donald.musictheoryapp.util.getDescriptions
import com.donald.musictheoryapp.util.getScore
import com.donald.musictheoryapp.music.musicxml.Note
import com.donald.musictheoryapp.util.getInputHint
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException

class IntervalInputQuestion(
    number: Int,
    descriptions: List<Description>,
    inputHint: String?,
    val score: Score,
    val requiredInterval: String,
    val answer: Answer
) : ChildQuestion(number, descriptions, inputHint) {

    override val points: Int
        get() = if (answer.correct) 1 else 0
    override val maxPoints = 1
    override val correctAnswerCount = 1

    override fun acceptVisitor(visitor: QuestionVisitor) {
        visitor.visit(this)
    }

    override fun copyNew(): IntervalInputQuestion {
        return IntervalInputQuestion(
            number,
            List(descriptions.size) { i -> descriptions[i].copy() },
            inputHint,
            score.copy(),
            requiredInterval,
            answer.copyNew()
        )
    }

    @Throws(JSONException::class)
    override fun toPartialJson(): JSONObject {
        val jsonObject = JSONObject()
        jsonObject.put(QUESTION_TYPE, Type.INTERVAL_INPUT.ordinal)
        jsonObject.put("score", score.toDocument().asXML())
        jsonObject.put("required_interval", requiredInterval)
        jsonObject.put("answer", answer.toJson())
        return jsonObject
    }

    companion object {

        @Throws(JSONException::class, IOException::class, XmlPullParserException::class)
        fun fromJson(jsonObject: JSONObject): IntervalInputQuestion {
            val question = IntervalInputQuestion(
                number = jsonObject.getInt("number"),
                descriptions = jsonObject.getDescriptions(),
                inputHint = jsonObject.getInputHint(),
                score = jsonObject.getScore(),
                requiredInterval = jsonObject.getString("required_interval"),
                answer = Answer.fromJson(jsonObject.getJSONObject("answer"))
            )
            check(!(question.score.parts().size != 1 || question.score.parts()[0].measures().size != 1 || question.score.parts()[0].measures()[0].notes().size != 1)) { "Invalid score from question." }
            return question
        }

    }

    class Answer(var userAnswer: Note?, val correctAnswer: Note) : ChildQuestion.Answer {

        override val correct: Boolean
            get() = userAnswer?.pitch?.equals(correctAnswer.pitch) ?: false

        override fun copyNew(): Answer {
            return Answer(null, correctAnswer)
        }

        @Throws(JSONException::class)
        fun toJson(): JSONObject {
            return JSONObject().apply {
                put("user_answer", userAnswer?.toJson() ?: JSONObject.NULL)
                put("correct_answer", correctAnswer.toJson())
            }
        }

        companion object {
            @Throws(JSONException::class)
            fun fromJson(jsonObject: JSONObject): Answer {
                return Answer(
                    if (jsonObject.isNull("user_answer")) null else Note.fromJson(jsonObject.getJSONObject("user_answer")),
                    Note.fromJson(jsonObject.getJSONObject("correct_answer"))
                )
            }
        }

    }

}